src/mbgl/map/source.cpp

```diff
@@ -603,4 +606,58 @@ void Source::dumpDebugLogs() const {
     }
 }
 
+FeatureResults Source::featuresAt(const PrecisionPoint point, const uint8_t radius, const TransformState& transform) const {
+    // figure out tile (bounded by source max zoom)
+    LatLng p = transform.pointToLatLng(point);
+
+    const double sine = std::sin(p.latitude * M_PI / 180);
+    const double x = p.longitude / 360 + 0.5;
+    double y = 0.5 - 0.25 * std::log((1 + sine) / (1 - sine)) / M_PI;
+
+    y = y < -1 ? -1 : y > 1 ? 1 : y;
+
+    const uint8_t z = ::floor(transform.getZoom());
+    const uint8_t source_max_z = ::fmin(z, info.max_zoom);
+    const uint32_t z2 = ::powf(2, source_max_z);
+
+    TileID id(z, ::floor(x * z2), ::floor(y * z2), source_max_z);
+
+    // figure out query bounds
+    TileCoordinate coordinate = transform.pointToCoordinate(point);
+    coordinate = coordinate.zoomTo(::fmin(id.z, info.max_zoom));
+
+    vec2<uint16_t> position((coordinate.column - id.x) * 4096, (coordinate.row - id.y) * 4096);
+
+    const uint32_t tile_scale = ::pow(2, id.z);
+    const double scale = util::tileSize * transform.getScale() / (tile_scale / id.overscaling);
+
+    const int16_t r = radius * 4096 / scale;
+
+    const int16_t left   = position.x - r;
+    const int16_t right  = position.x + r;
+    const int16_t bottom = position.y - r;
+    const int16_t top    = position.y + r;
+
+    FeatureBox queryBox = {{ left, bottom }, { right, top }};
+
+    FeatureResults results;
+
+    // find the right tile
+    for (const auto& tile : getLoadedTiles()) {
```

Comment:
This seems really inefficient; can we instead implement a function that just gets the correct tile from the unordered map?